pylint/checkers/unicode.py

```diff
@@ -0,0 +1,670 @@
+            f'use "{self.escaped}" instead.'
+        )
+
+    def human_code(self) -> str:
+        """used to generate the human readable error message"""
+        return f"invalid-ascii-{self.name}"
+
+
+def map_positions_to_result(
+    line: StrLike,
+    search_dict: Dict[StrLike, BadASCIIChar],
+) -> Dict[int, BadASCIIChar]:
+    """Only required due to typing issue"""
+
+    BYTES_NEWLINE = 10  # b"\n"  - If taking some_byte[-1], the result is an integer
+
+    result: Dict[int, BadASCIIChar] = {}
+    ignore_pos: Optional[int]
+    for search_for, char in search_dict.items():
+        # Special Handling for Windows '\r\n'
+        if char.unescaped == "\r" and line[-1] in ("\n", BYTES_NEWLINE):
+            ignore_pos = len(line) - 2
+        else:
+            ignore_pos = None
+
+        start: int = 0
+        pos: int = line.find(search_for, start)
+        while pos > 0:
+            if pos != ignore_pos:
+                result[pos] = char
+            start = pos + 1
+            pos = line.find(search_for, start)
+
+    return result
+
+
+@lru_cache(maxsize=1)
+def get_bom_sorted() -> Dict[bytes, str]:
+    """
+    Get all BOM sorted by size, biggest first
+    """
+    boms_unsorted = {
+        codecs.BOM_UTF8: "utf-8",
+        codecs.BOM_UTF16_LE: "utf-16le",
+        codecs.BOM_UTF16_BE: "utf-16be",
+        codecs.BOM_UTF32_LE: "utf-32le",
+        codecs.BOM_UTF32_BE: "utf-32be",
+    }
+    result = OrderedDict()
+    for key in sorted(boms_unsorted.keys(), key=len, reverse=True):
+        result[key] = boms_unsorted[key]
+    return result
+
+
+@lru_cache(maxsize=1000)
+def cached_encode(string: str, encoding: str) -> bytes:
+    """
+    A cached version of encode
+    """
+    return string.encode(encoding)
+
+
+def extract_codec_from_bom(first_line: bytes) -> Optional[str]:
+    """
+    Try to extract the codec (unicode only) by checking for the BOM
+
+    see https://unicode.org/faq/utf_bom.html#BOM
+
+    Default to UTF-8 if no BOM is found
+
+    Args:
+        first_line: the first line of a file
+
+    Returns:
+        a codec name
+    """
+    for bom, codec in get_bom_sorted().items():
+        if first_line.startswith(bom):
+            return codec
+
+    return None
+
+
+class InvalidUnicodeModeChecker(BaseChecker):
+    """
+    Check that UTF-16 and UTF-32 are not used to encode Python Files.
+
+    Currently Python supports only UTF-8. See
+    https://stackoverflow.com/questions/69897842/ and https://bugs.python.org/issue1503789
+    for background.
+    This simplifies the BiDirectional Unicode check below as we only need to check
+    for UTF-8 bytes.
+
+    As this is a security related check we want to ensure that even once Python
+    support UTF-16 and UTF-32 it is not possible for an attacker to bypass our checks
+    by simply using a new encoding.
+    """
+
+    __implements__ = interfaces.IRawChecker
+
+    msgs = {
+        "E1800": (
+            "Use only UTF-8, not UTF-16 or UTF-32!",
```

Comment:
```suggestion
            "Only UTF-8 encoding is supported by Python",
```

Reply:
Actually that is not true.
Currently Python supports all codecs besides UTF-16 and UTF-32, if the codec is defined as described in [PEP263](https://www.python.org/dev/peps/pep-0263/).
If a users sees this error, this means the AST Parser is not crashing anymore, which would mean that Python version used indeed would then supports UTF-16 and UTf-32: This case is what what the check is for.

But seeing your comments I see that I didn't document this very well :-/

Reply:
Maybe we should write: "UTF-8 is recommended encoding for Python files. Don't use UTF-16 or UTF-32."

Reply:
Hm, I misread this then yeah. But then perhaps we shouldn't make this an `Error`? But `Refactoring` like many of the other `consider-...` checks? If it is not actually invalid the message name doesn't really reflect that right now.

Reply:
The problem is, that many people turn off all Refactoring Messages (at least in our company).
The Idea behind all this checkers is that we find possible malicious code.

As UTF-16/UTF-32 would circumstance our protections, that would be an issues.

Maybe we can directly tell the user what this is all about:
"UTF-16 nor UTF-32 are supported by Pylints Unicode checker. Use UTF-8 instead!",
"As of writing the Pylint Unicode checker, Python didn't support UTF-16/UTF-32. Therefore I won't work properly with these encodings."

Is this understandable for you?

Reply:
Due to combining the checkers I won't have to assume UTF-8 anymore. I will add some more future proof tests and Change the message o a Convention or Refactoring check (will look in the examples before).

Reply:
Ok I had to invest a shit lot of work into getting it stuff to work with UTF-16/UTF-32.
There are a lot of edge cases that people can run into, when using Python with UTF-16 even if it would be ever supported.

So I will still argue that Pylint should raise an error instead of a Convention.
I adopted the error message.

If you want something else @DanielNoord let me know.